WordPressKit/Domains/DomainsServiceRemote+AllDomains.swift

```diff
@@ -0,0 +1,182 @@
+import Foundation
+
+extension DomainsServiceRemote {
+
+    // MARK: - API
+
+    /// Makes a call request to `GET /v1.1/all-domains` and returns a list of domain objects.
+    ///
+    /// The endpoint accepts 3 **optionals** query params:
+    /// - `resolve_status` of type `boolean`. If `true`, the response will include a `status` attribute for each `domain` object.
+    /// - `no_wpcom`of type `boolean`. If `true`, the respnse won't include `wpcom` domains.
+    /// - `locale` of type `string`. Used for string localization.
+    public func fetchAllDomains(params: AllDomainsEndpointParams? = nil, completion: @escaping (AllDomainsEndpointResult) -> Void) {
+        let path = self.path(forEndpoint: "all-domains", withVersion: ._1_1)
+        var parameters: [String: AnyObject]?
+
+        do {
+            parameters = try queryParameters(from: params)
+        } catch let error {
+            completion(.failure(error))
+            return
+        }
+
+        self.wordPressComRestApi.GET(path, parameters: parameters) { result, _ in
+            do {
+                switch result {
+                case .success(let result):
+                    let data = try JSONSerialization.data(withJSONObject: result, options: [])
+                    let decoder = JSONDecoder()
+                    decoder.dateDecodingStrategy = .iso8601
+                    let result = try decoder.decode(AllDomainsEndpointResponse.self, from: data)
+                    completion(.success(result.domains))
+                case .failure(let error):
+                    throw error
+                }
+            } catch let error {
+                completion(.failure(error))
+            }
+        }
+    }
+
+    private func queryParameters(from params: AllDomainsEndpointParams?) throws -> [String: AnyObject]? {
+        guard let params else {
+            return nil
+        }
+        let encoder = JSONEncoder()
+        let data = try encoder.encode(params)
+        let dict = try JSONSerialization.jsonObject(with: data) as? [String: AnyObject]
+        return dict
+    }
+
+    // MARK: - Public Types
+
+    public typealias AllDomainsEndpointResult = Result<[AllDomainsListItem], Error>
+
+    public struct AllDomainsEndpointParams {
+
+        public var resolveStatus: Bool = false
+        public var noWPCOM: Bool = false
+        public var locale: String?
+
+        public init() {}
+    }
```

Comment:
[Suggestion] Have you considered doing something like this?
```suggestion
    public struct AllDomainsEndpointParams {

        let resolveStatus: Bool
        let noWPCOM: Bool
        let locale: String?

        public init(resolveStatus: Bool = false, noWPCOM: Bool = false, locale: String?) {
            self.resolveStatus = resolveStatus
            self.noWPCOM = noWPCOM
            self.locale = locale
        }
    }
```

Reply:
Yes, but I wasn't sure about rewriting those properties in the init. Consider a situation where the `struct` had +10 properties, then the init would have the same +10 properties, In my opinion, that's redundant. 

In UIKit and Foundation, there are many `structs` that you can mutate their properties directly. So I think the practice of mutable struct properties is not uncommon.

If you feel strongly about this, I can open another PR and make the change.

Reply:
> In UIKit and Foundation, there are many `structs` that you can mutate their properties directly. So I think the practice of mutable struct properties is not uncommon.

Thank you for your rationale @salimbraksa 😄 

I feel strongly about this, actually. I shall P2 about this to make my case. It would be good to bubble up and have an agreed upon guideline. In the meantime, as far as I'm concerned, the implementation can stay like this. 👍